s3.go

```diff
@@ -1,211 +1,313 @@
-			tok, err := dec.Token()
+		for index >= len(resp.Contents) {
+			if !*resp.IsTruncated {
+				return dsq.Result{}, false
+			}
+
+			resp, err = s.S3.ListObjectsV2(&s3.ListObjectsV2Input{
+				Bucket:            aws.String(s.Bucket),
+				Prefix:            aws.String(s.s3Path(q.Prefix)),
+				Delimiter:         aws.String("/"),
+				MaxKeys:           aws.Int64(listMax),
+				ContinuationToken: resp.NextContinuationToken,
+			})
 			if err != nil {
-				if err == io.EOF {
-					return dsq.Result{}, false
-				}
 				return dsq.Result{Error: err}, false
 			}
 
-			switch tok := tok.(type) {
-			case xml.StartElement:
-				if tok.Name.Local == "Key" {
-					nextiskey = true
-				}
-			case xml.CharData:
-				if nextiskey {
-					kval := string(tok)
-					nextiskey = false
-					dec.Skip()
-
-					if strings.HasPrefix(kval, ".md5/") {
-						continue
-					}
-					res := dsq.Result{Entry: dsq.Entry{Key: kval}}
-					return res, true
-				}
+			index -= len(resp.Contents)
+			if index < 0 {
+				index = 0
 			}
 		}
+
+		entry := dsq.Entry{
+			Key: *resp.Contents[index].Key,
+		}
+		if !q.KeysOnly {
+			value, err := s.Get(ds.NewKey(entry.Key))
+			if err != nil {
+				return dsq.Result{Error: err}, false
+			}
+			entry.Value = value
+		}
+
+		index++
+		return dsq.Result{Entry: entry}, true
 	}
 
 	return dsq.ResultsFromIterator(q, dsq.Iterator{
 		Close: func() error {
-			return resp.Body.Close()
+			return nil
 		},
 		Next: nextValue,
 	}), nil
 }
 
+func (s *S3Bucket) Batch() (ds.Batch, error) {
+	return &s3Batch{
+		s:   s,
+		ops: make(map[string]batchOp),
+	}, nil
+}
+
 func (s *S3Bucket) Close() error {
+	close(s.limiter)
 	return nil
 }
 
+func (s *S3Bucket) s3Path(p string) string {
+	return path.Join(s.RootDirectory, p)
+}
+
+func parseError(err error) error {
+	if s3Err, ok := err.(awserr.Error); ok && s3Err.Code() == s3.ErrCodeNoSuchKey {
+		return ds.ErrNotFound
+	}
+	return err
+}
+
 type s3Batch struct {
-	s *S3Bucket
+	s   *S3Bucket
+	ops map[string]batchOp
 }
 
-func (s *S3Bucket) Batch() (ds.Batch, error) {
-	return &s3Batch{s}, nil
+type batchOp struct {
+	val    []byte
+	delete bool
 }
 
 func (b *s3Batch) Put(k ds.Key, val []byte) error {
-	return b.s.Put(k, val)
+	b.ops[k.String()] = batchOp{
+		val:    val,
+		delete: false,
+	}
+	return nil
 }
 
 func (b *s3Batch) Delete(k ds.Key) error {
-	return b.s.Delete(k)
+	b.ops[k.String()] = batchOp{
+		val:    nil,
+		delete: true,
+	}
+	return nil
 }
 
 func (b *s3Batch) Commit() error {
+	var (
+		deleteObjs []*s3.ObjectIdentifier
+		putKeys    []ds.Key
+	)
+	for k, op := range b.ops {
+		if op.delete {
+			deleteObjs = append(deleteObjs, &s3.ObjectIdentifier{
+				Key: aws.String(k),
+			})
+		} else {
+			putKeys = append(putKeys, ds.NewKey(k))
+		}
+	}
+
+	errChanSize := len(putKeys)
+	if len(deleteObjs) > 0 {
+		errChanSize++
+	}
+	errChan := make(chan error, errChanSize)
+	defer close(errChan)
+
+	for _, k := range putKeys {
+		go func(k ds.Key, op batchOp) {
```

Comment:
This is going to launch quite a few goroutines. Unfortunately, this isn't as "free" as the go documentation would lead you to believe. I'd *prefer* to spawn N goroutines and then feed them with a channel.

Reply:
As in it will take up memory on the heap? I'm okay with switching the concurrency pattern.

Reply:
That and it'll cost the scheduler something. I mean, really, it's probably not that big of an issue. There are just better ways to do this that aren't *that* complicated.